tests/specs/infinite-gas.k

```diff
@@ -44,19 +52,34 @@ module INFINITE-GAS-JAVA [kast]
     rule I1         <=Int I2 +Int I3 => I1 -Int I3 <=Int I2         requires #isConcrete(I1) andBool #isConcrete(I3) [simplification]
 
-    rule ((I1 +Int I2) +Int I3) +Int I4 => (I1 +Int I3) +Int (I2 +Int I4) requires #isConcrete(I2) andBool #isConcrete(I4) andBool notBool (#isConcrete(I1) orBool #isConcrete(I3)) [simplification]
-    rule ((I1 +Int I2) -Int I3) +Int I4 => (I1 -Int I3) +Int (I2 +Int I4) requires #isConcrete(I2) andBool #isConcrete(I4) andBool notBool (#isConcrete(I1) orBool #isConcrete(I3)) [simplification]
-    rule (I1 +Int I2) -Int (I3 +Int I4) => (I1 -Int I3) +Int (I2 -Int I4) requires #isConcrete(I2) andBool #isConcrete(I4) andBool notBool (#isConcrete(I1) orBool #isConcrete(I3)) [simplification]
+    rule I1 +Int I2 => I2 +Int I1 requires #isConcrete(I1) andBool notBool #isConcrete(I2) [simplification]
+
+    rule (I1 +Int I2) +Int I3 => (I1 +Int I3) +Int I2 requires #isConcrete(I2) andBool notBool #isConcrete(I3) [simplification]
+    rule (I1 +Int I2) -Int I3 => (I1 -Int I3) +Int I2 requires #isConcrete(I2) andBool notBool #isConcrete(I3) [simplification]
+    rule (I1 -Int I2) +Int I3 => (I1 +Int I3) -Int I2 requires #isConcrete(I2) andBool notBool #isConcrete(I3) [simplification]
+    rule (I1 -Int I2) -Int I3 => (I1 -Int I3) -Int I2 requires #isConcrete(I2) andBool notBool #isConcrete(I3) [simplification]
+
+    rule (I1 +Int I2) +Int I3 => (I2 +Int I3)          +Int I1 requires #isConcrete(I2) andBool #isConcrete(I3) andBool notBool #isConcrete(I1) [simplification]
+    rule (I1 +Int I2) -Int I3 => (I2 -Int I3)          +Int I1 requires #isConcrete(I2) andBool #isConcrete(I3) andBool notBool #isConcrete(I1) [simplification]
+    rule (I1 -Int I2) +Int I3 => (I3 -Int I2)          +Int I1 requires #isConcrete(I2) andBool #isConcrete(I3) andBool notBool #isConcrete(I1) [simplification]
+    rule (I1 -Int I2) -Int I3 => ((0 -Int I2) -Int I3) +Int I1 requires #isConcrete(I2) andBool #isConcrete(I3) andBool notBool #isConcrete(I1) [simplification]
 endmodule
 
 module INFINITE-GAS-COMMON
     imports EVM
 
  // Generic rules
 
-    rule I -Int I => 0                                              [simplification]
-    rule I1 -Int (I1 -Int I2) => I2                                 [simplification]
-    rule (notBool (A andBool B)) andBool A => (notBool B) andBool A [simplification]
+    rule (notBool (A andBool B)) andBool A => (notBool B) andBool A                      [simplification]
+    rule #if B #then C +Int C1 #else C +Int C2 #fi => C +Int #if B #then C1 #else C2 #fi [simplification]
+
+    rule I -Int I => 0              [simplification]
+    rule (I1 -Int I2) +Int I2 => I1 [simplification]
+
+    rule I1 +Int (I2 +Int I3) => (I1 +Int I2) +Int I3 [simplification]
+    rule I1 +Int (I2 -Int I3) => (I1 +Int I2) -Int I3 [simplification]
+    rule I1 -Int (I2 +Int I3) => (I1 -Int I2) -Int I3 [simplification]
+    rule I1 -Int (I2 -Int I3) => (I1 -Int I2) +Int I3 [simplification]
```

Comment:
If there wouldn't be rule priorities, this would still lead to infinite loop in combination with rule on L29.

In practice infinite loop probably won't happen, but for rather peculiar reasons. In Haskell backend definition rules are applied before `[simplification]`, so concrete `(I2+I3)` will evaluate first. in Java although it's not documented, builtin evaluation happens before rules are applied.

So I suggest either the following comment:
```
Without rule priorities, these rules could lead to infinite loop in combination with rule
(add a label to lemma on L29). However, Haskell backend executes definition rules 
before simplification rules, 
and Java backend executes builtin evaluations like concrete `+Int` before any rules.
```

Or alternatively, you'll have to duplicate the rules above for Java and Haskell and frame then with some concrete/symbolic conditions.

Reply:
Can you come up with an example where this infinite loops with the rules on lines 29/55? I can't see it.

Reply:
`(1+1)+A => A+(1+1) (L29) => (A+1)+1 (L79) => (1+1)+A (L36)`

Besides, we have a version of such rules in verified smart contracts repo. There I used this naming convention for clarity. Concrete vars are called I1, I2 etc, symbolic vars A1, A2, ...

Reply:
Thanks for the example. I reviewed this example with teh updated code, and while it looks like there are multiple execution paths, none of them loop and they all arrive at the same answer.